### PR TITLE
Feature/tao 4465 runner optimizations

### PR DESCRIPTION
**Pull request by krampstudio**
### Structural changes
 - Removal of the _submitItem_ step in most of the cases.
 - We check if the item will have _modalFeedbacks_ upfront and use the _submitItem_ step only if the item has feedback.
 - unified HTTP parameters for move, skip, timeout, exit and _submitItem_.
 - the _allowSkipping_ is now implemented client side (using the `preventSkipping` plugin) with fallback validation server side (the _submitItem_ request was used also for that).
 - rubric blocks where previously in the _getItem_ response, they are now part of the `testContext` (so we don't open the test session in the _getItem_ query).
 - move the function `isLeavingSection` to a new helper `taoQtiTest/runner/helpers/navigation`.

### Modifications in the HTTP requests

- `/taoQtiTest/Runner/submitItem` isn't used any more expect in the specific case of modal feedback.
- `/taoQtiTest/Runner/move`, `/taoQtiTest/Runner/timeout`, `/taoQtiTest/Runner/submitItem` and `/taoQtiTest/Runner/exitTest` have now all the following additional parameters :
    -  `itemState`  : A JSON string that contains the item state to save
    -  `itemResponses` : A JSON string that contains the item responses
    -  `itemIdentifier` : the identifier of the item
    -  `itemDuration` : the duration of the item is seconds (a float)
    -  `consumedExtraTime` :   extra time is seconds (a float)
- `/taoQtiTest/Runner/getItem` has now an additional parameter
    -  `itemIdentifier` : the identifier of the item
- `/taoQtiTest/Runner/getItem` has lost  the `rubricBlock` from the response
- All requests that produce a `testContext` (`init`, `move`, etc.), have now the following additional entries in the response :
    - `testContext.preventEmptyResponses=<true|false>`
    - `testContext.hasFeedbacks=<true|false>`
    - `testContext.rubrics=<undefined|string|string[]>`

### CODE REVIEW

File: views/js/runner/plugins/navigation/preventSkipping.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                .after('move skip', function(){
+                    toggle();
+                });
+        },
+
+        /**
+         * Called during the runner's destroy phase
+         */
+        destroy : function destroy (){
+        },
```
Comment (jsconan) [inline views/js/runner/plugins/navigation/preventSkipping.js]: Is this really needed ?

File: views/js/runner/provider/qti.js [hunk 2]
```diff
@@ -190,118 +208,103 @@ define([
...
+
+                                    if(results.displayFeedbacks === true && results.feedbacks) {
+                                        self.itemRunner.renderFeedbacks(results.feedbacks, results.itemSession, function(queue){
+                                            self.trigger('modalFeedbacks', queue, resolve);
+                                        });
+                                        return;
+                                    }
+                                }
+                                return resolve();
+                            });
```
Comment (jsconan) [inline views/js/runner/provider/qti.js]: There is no `catch` on this promise. What happen if the server refuse the `submit` ?

File: views/js/runner/provider/qti.js [hunk 3]
```diff
@@ -190,118 +208,103 @@ define([
...
-                                        self.trigger('modalFeedbacks', queue, resolve);
-                                    });
+                                if (results.itemSession) {
+                                    context.itemAnswered = results.itemSession.itemAnswered;
 
+                                    if(results.displayFeedbacks === true && results.feedbacks) {
+                                        self.itemRunner.renderFeedbacks(results.feedbacks, results.itemSession, function(queue){
+                                            self.trigger('modalFeedbacks', queue, function(){
+                                                computeNext(action, _.omit(params, 'itemState', 'itemResponse'));
+                                            });
```
Comment (jsconan) [inline views/js/runner/provider/qti.js]: Maybe I am wrong, but is there chance the feedbacks be displayed twice? There is two places where the feedbacks are handled. I suppose they are sent by the server only after a response submit?

File: views/js/runner/provider/qti.js [hunk 4]
```diff
@@ -430,20 +390,12 @@ define([
...
-                            });
-                        })
-                        .catch(function(err){
-                            self.trigger('error', err);
-                        });
+                    self.trigger('alert.timeout', __('Time limit reached, this part of the test has ended.'), function() {
+                        computeNext('timeout', _.merge(getItemResults(), {
+                            scope: scope,
+                            ref: ref
+                        }));
```
Comment (jsconan) [inline views/js/runner/provider/qti.js]: I see a potential issue here, as you display the timeout alert before requesting the server.
What happen if the test taker takes a while before acknowledging the message? Or just close the browser, and open the test on another one?
Maybe should it be better to request the server, and wait for the test taker before displaying the result of the request, should it not?

File: actions/class.Runner.php [hunk 5]
```diff
@@ -662,7 +688,18 @@ public function timeout()
...
+
+            if (!$this->runnerService->isTerminated($serviceContext)) {
+                $this->endItemTimer($serviceContext);
+                $this->saveItemState($serviceContext);
+            }
+
+            $this->runnerService->check($serviceContext);
+            $this->serviceContext->init();
+
+            $this->saveItemResponses($serviceContext);
```
Comment (krampstudio) [inline actions/class.Runner.php]: `saveItemResponses` is called in different manner, but I'll factorize the context init/check

File: views/js/runner/provider/qti.js (hunk 3 above)
Comment (krampstudio) [inline views/js/runner/provider/qti.js]: I just forgot to remove one of them

File: views/js/runner/provider/qti.js (hunk 4 above)
Comment (krampstudio) [inline views/js/runner/provider/qti.js]: just pushed a fixed were the timeout request is pushed before displaying the message

**Comment by krampstudio** [comment]
I'm wondering if such HTTP changes aren't considered as breaking changes in semver

**Comment by jsconan** [comment]
Yes, I had the same wondering.

**Pull request merged**